README.md
# eloquent-queryable

[![Latest Version on Packagist][ico-version]][link-packagist]
[![Software License][ico-license]](LICENSE.md)
[![Build Status][ico-travis]][link-travis]
[![Coverage Status][ico-scrutinizer]][link-scrutinizer]
[![Quality Score][ico-code-quality]][link-code-quality]
[![Total Downloads][ico-downloads]][link-downloads]

A PHP package to automatically inject query rules in Eloquent model queries from request query parameters.

Filtering query results in Laravel/Lumen application with Eloquent is a very common task that all the developers need to do in almost any project. The purpose of this package is to provide a single trait to give the ability to filter and add query logic to Elouqent models using request query strings.

For example, the folling query will fetch all the users whos names starts with B, order them by created at and will also fetch only non admin users.

```
http://localhost:8000/users?q[where_name_starts_with]=b&q[order_by]=created_at,asc&q[where_is_admin_ne]=1
```


## Install

**Via Composer**

``` bash
$ composer require munza/eloquent-queryable
```

## Register

**in Laravel**

```php
// config/app.php

    'providers' => [
    // ...
        Munza\EloquentQueryable\EloquentQueryableServiceProvider::class,
    // ...
    ]
```

**in Lumen**

```php
// bootstrap/app.php

    // ...
    $app->register(Munza\EloquentQueryable\EloquentQueryableServiceProvider::class);
    // ...
```

## Configuration

**in Laravel**

```bash
$ php artisan vendor:publish --provider=Munza\EloquentQueryable\EloquentQueryableServiceProvider::class
```

**in Lumen**

```bash
$ cp vendor/munza/eloquent-queryable/resources/config/eloquent-queryable.php ./config
```

```php
// bootstrap/app.php
// ...
$app->configure('eloquent-queryable');
// ...
```

## Usage

Add `Munza\EloquentQueryable\Traits\RequestQueryable` trait in the model to use.

## Example Queries

```
http://localhost:8000/users?q[select]=id,name
http://localhost:8000/users?q[order_by]=name,desc;created_at
http://localhost:8000/users?q[limit]=20
http://localhost:8000/users?q[offset]=10
http://localhost:8000/users?q[with]=posts,posts.comments

http://localhost:8000/users?q[where_{column_name}_{postfix}]={values}
http://localhost:8000/users?q[or_where_{column_name}_{postfix}]={values}

http://localhost:8000/users?q[where_name_starts_with]=John
http://localhost:8000/users?q[where_name_not_starts_with]=John

http://localhost:8000/users?q[where_name_ends_with]=Doe
http://localhost:8000/users?q[where_name_not_ends_with]=Doe

http://localhost:8000/users?q[where_name_contains]=John
http://localhost:8000/users?q[where_name_not_contains]=John

http://localhost:8000/users?q[where_created_at_gt]=2017-07-19%2008:14:45
http://localhost:8000/users?q[where_created_at_lt]=2017-07-19%2008:14:48

http://localhost:8000/users?q[where_name_eq]=John%20Doe
http://localhost:8000/users?q[where_name_ne]=John%20Doe

http://localhost:8000/users?q[where_id_in]=1,2,3
http://localhost:8000/users?q[where_id_not_in]=1,2,3

http://localhost:8000/users?q[where_created_at_between]=2017-07-19%2008:14:45,2017-07-19%2008:14:48
http://localhost:8000/users?q[where_created_at_not_between]=2017-07-19%2008:14:45,2017-07-19%2008:14:48
```

## Change log

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Testing

``` bash
$ composer test
```

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) and [CONDUCT](CONDUCT.md) for details.

## Security

If you discover any security related issues, please email tawsif.aqib@gmail.com instead of using the issue tracker.

## Credits

- [Tawsif Aqib][link-author]
- [All Contributors][link-contributors]

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

[ico-version]: https://img.shields.io/packagist/v/munza/eloquent-queryable.svg?style=flat-square
[ico-license]: https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square
[ico-travis]: https://img.shields.io/travis/munza/eloquent-queryable/master.svg?style=flat-square
[ico-scrutinizer]: https://img.shields.io/scrutinizer/coverage/g/munza/eloquent-queryable.svg?style=flat-square
[ico-code-quality]: https://img.shields.io/scrutinizer/g/munza/eloquent-queryable.svg?style=flat-square
[ico-downloads]: https://img.shields.io/packagist/dt/munza/eloquent-queryable.svg?style=flat-square

[link-packagist]: https://packagist.org/packages/munza/eloquent-queryable
[link-travis]: https://travis-ci.org/munza/eloquent-queryable
[link-scrutinizer]: https://scrutinizer-ci.com/g/munza/eloquent-queryable/code-structure
[link-code-quality]: https://scrutinizer-ci.com/g/munza/eloquent-queryable
[link-downloads]: https://packagist.org/packages/munza/eloquent-queryable
[link-author]: https://munza.github.io
[link-contributors]: ../../contributors


CHANGELOG.md
# Changelog

All Notable changes to `eloquent-queryable` will be documented in this file.

Updates should follow the [Keep a CHANGELOG](http://keepachangelog.com/) principles.

## NEXT - YYYY-MM-DD

### Added
- Nothing

### Deprecated
- Nothing

### Fixed
- Nothing

### Removed
- Nothing

### Security
- Nothing


src/Traits/RequestQueryParam.php
<?php

namespace Munza\EloquentQueryable\Traits;

trait RequestQueryable
{
    /**
     * {@inheritdoc}
     */
    public function newQuery()
    {
        $queryParams = $this->getQueryParams();
        $newQuery = parent::newQuery();

        foreach ($queryParams as $queryParam => $args) {
            foreach(
                config('eloquent-queryable.query_keys.prefix') as $key => $value
            ) {
                if (starts_with($queryParam, $value)) {
                    $this->addQueryCondition(
                        $newQuery, $queryParam, $args, $key
                    );
                    break;
                }
            }
        }

        return $newQuery;
    }

    /**
     * Add a basic query condition to the query builder.
     *
     * @param \Illuminate\Database\Eloquent\Builder $query
     * @param array                                 $queryParam
     * @param array                                 $args
     * @param string                                $queryKey
     */
    private function addQueryCondition($query, $queryParam, $args, $queryKey)
    {
        switch ($queryKey) {
            case 'where':
                $this->addWhereQuery($query, $queryParam, $args);
                break;
            case 'or_where':
                $this->addWhereQuery($query, $queryParam, $args, true);
                break;
            default:
                foreach ($args as $arg) {
                    $query->{$queryKey}(...$arg);
                }
                break;
        }
    }

    /**
     * Get the 'q' query string from the url
     * and return a formatted array.
     *
     * @return array
     */
    private function getQueryParams()
    {
        $queryParams = app('request')->get('q');

        if (is_null($queryParams)) return [];

        foreach ($queryParams as $queryParam => $value) {
            $queryParams[$queryParam] = $this->parseQueryArgs($value);
        }

        return $queryParams;
    }

    /**
     * Parse the query argument string
     * and convert into an array.
     *
     * @param  string $arg
     * @return array
     */
    private function parseQueryArgs($arg)
    {
        return array_map(function($item) {
            return explode(',', $item);
        }, explode(';', $arg));
    }

    /**
     * Get the where query condition from the query params.
     *
     * @param  array $queryParam
     * @return array
     */
    private function getWhereQueryCondition($queryParam, $or = false)
    {
        $whereType = $or
            ? config('eloquent-queryable.query_keys.prefix.or_where')
            : config('eloquent-queryable.query_keys.prefix.where');

        list($match, $comparator, $wrapper) = [null, null, null];

        foreach (
            config('eloquent-queryable.query_keys.postfix')as $key => $value
        ) {
            if (ends_with($queryParam, $value)) {
                preg_match(
                    "/^{$whereType}_(.*?)_{$value}$/", $queryParam, $match
                );
                list($comparator, $wrapper) =
                    $this->getComparatorAndWrapperFor($key);
                break;
            }
        }

        return [$match ? $match[1] : null, $comparator, $wrapper];
    }

    /**
     * Get the comparator and wrapper for a postfix key.
     *
     * @param  string $key
     * @return array
     */
    private function getComparatorAndWrapperFor($key)
    {
        $key_to_comparator_wrapper_map = [
            'not_starts_with' => ['not like', ['', '%']],
            'starts_with'     => ['like',     ['', '%']],
            'not_ends_with'   => ['not like', ['%', '']],
            'ends_with'       => ['like',     ['%', '']],
            'not_contains'    => ['not like', ['%', '%']],
            'contains'        => ['like',     ['%', '%']],
            'greater_than'    => ['>',        ['', '']],
            'lower_than'      => ['<',        ['', '']],
            'equals_to'       => ['=',        ['', '']],
            'not_equal'       => ['<>',       ['', '']],
            'not_in'          => [null,       null],
            'in'              => [null,       null],
            'not_between'     => [null,       null],
            'between'         => [null,       null],
        ];

        return array_key_exists($key, $key_to_comparator_wrapper_map)
            ? $key_to_comparator_wrapper_map[$key]
            : [null, null];
    }

    /**
     * Add where condition in the query pipeline.
     *
     * @param \Illuminate\Database\Eloquent\Builder $query
     * @param array                                 $queryParam
     * @param array                                 $args
     * @param bool                                  $or
     */
    private function addWhereQuery($query, $queryParam, $args, $or = false)
    {
        list($field, $comparator, $argsWrapper) =
            $this->getWhereQueryCondition($queryParam, $or);

        // For where query conditions with comparator
        if ($field && $comparator && is_array($argsWrapper)) {
            $query->{$or ? 'orWhere' : 'where'}(
                $field,
                $comparator,
                $argsWrapper[0].$args[0][0].$argsWrapper[1]
            );
        }

        // For where in and between query conditions
        if ($field && !$comparator && !is_array($argsWrapper)) {
            foreach ([
                'not_in',
                'in',
                'not_between',
                'between',
            ] as $key) {
                if (ends_with($queryParam, $key)) {
                    $where = ($or ? 'orWhere' : 'where')
                        .(starts_with($key, 'not_') ? 'Not' : '')
                        .studly_case(str_replace('not_', '', $key));

                    $query->{$where}($field, $args[0]);
                    break;
                }
            }
        }
    }

}


src/EloquentQueryableServiceProvider.php
<?php

namespace Munza\EloquentQueryable;

use Illuminate\Support\ServiceProvider;

class EloquentQueryableServiceProvider extends ServiceProvider
{
    /**
     * Perform post-registration booting of services.
     *
     * @return void
     */
    public function boot()
    {
        $this->publishes([
            __DIR__.'/../resources/config/eloquent-queryable.php'
                => base_path('config/eloquent-queryable.php'),
        ], 'config');
    }

    /**
     * Register any package services.
     *
     * @return void
     */
    public function register()
    {
        if (config('eloquent-queryable')) {
            $this->mergeConfigFrom(
                __DIR__.'/../resources/config/eloquent-queryable.php',
                'eloquent-queryable'
            );
        }
    }
}


resources/config/eloquent-queryable.php
<?php

return [

    'query_keys' => [

        /*
        |--------------------------------------------------------------------------
        | Basic Query Prefix
        |--------------------------------------------------------------------------
        |
        | eg. ?q[select]=id,name
        |     ?q[order_by]=name,desc;created_at
        |     ?q[limit]=20
        |     ?q[offset]=10
        |     ?q[with]=posts,posts.comments
        |     ?q[where_{column_name}_{postfix}]={values}
        |     ?q[or_where_{column_name}_{postfix}]={values}
        */
        'prefix' => [
            'select'   => 'select',
            'order_by' => 'order_by',
            'limit'    => 'limit',
            'offset'   => 'offset',
            'with'     => 'with',
            'where'    => 'where',
            'or_where' => 'or_where',
        ],

        /*
        |--------------------------------------------------------------------------
        | Where Query Postfix
        |--------------------------------------------------------------------------
        |
        | Only usable for 'where' and 'or_where' prefix.
        |
        | eg. ?q[where_name_starts_with]=John
        |     ?q[where_name_not_starts_with]=John
        |     ?q[where_name_ends_with]=Doe
        |     ?q[where_name_not_ends_with]=Doe
        |     ?q[where_name_contains]=John
        |     ?q[where_name_not_contains]=John
        |     ?q[where_created_at_gt]=2017-07-19%2008:14:45
        |     ?q[where_created_at_lt]=2017-07-19%2008:14:48
        |     ?q[where_name_eq]=John%20Doe
        |     ?q[where_name_ne]=John%20Doe
        |     ?q[where_id_in]=1,2,3
        |     ?q[where_id_not_in]=1,2,3
        |     ?q[where_created_at_between]=2017-07-19%2008:14:45,2017-07-19%2008:14:48
        |     ?q[where_created_at_not_between]=2017-07-19%2008:14:45,2017-07-19%2008:14:48
        */
        'postfix' => [
            'not_starts_with' => 'not_starts_with',
            'starts_with'     => 'starts_with',
            'not_ends_with'   => 'not_ends_with',
            'ends_with'       => 'ends_with',
            'not_contains'    => 'not_contains',
            'contains'        => 'contains',
            'greater_than'    => 'gt',
            'lower_than'      => 'lt',
            'equals_to'       => 'eq',
            'not_equal'       => 'ne',
            'where_not_in'    => 'not_in',
            'where_in'        => 'in',
            'not_between'     => 'not_between',
            'between'         => 'between',
        ],

    ],

];
